In [49]:
import numpy as np
import scipy as sp
from scipy.optimize import brentq

In [50]:
pv = 100000
fv = 0

In [51]:
pmt1 = -1500
n1 = 120

In [52]:
pmt2 = -1250
n2 = 180

In [53]:
pmt3 = -1160
n3 = 240

In [54]:
def fun1(i):
    return pv*(1+i)**n1 + pmt1*(((1+i)**n1-1)/i) + fv 
def fun2(i):
    return pv*(1+i)**n2 + pmt2*(((1+i)**n2-1)/i) + fv 
def fun3(i):
    return pv*(1+i)**n3 + pmt3*(((1+i)**n3-1)/i) + fv 

In [55]:
sol1 = brentq(fun1, 0.01, 1.0)
sol1

0.010930603939868333

In [56]:
sol2 = brentq(fun2, 0.01, 1.0)
sol2

0.01063980717924384

In [57]:
sol3 = brentq(fun3, 0.01, 1.0)
sol3

0.010697698792848072

In [58]:
def solve_fun(pv, pmt, n, fv):
    def fun(i):
        return pv*(1+i)**n + pmt*((((1+i)**n)-1)/i) + fv
    return brentq(fun, 0.001, 1.0)

In [59]:
i1 = solve_fun(pv, pmt1, n1, fv)
print(sol1)
i1

0.010930603939868333


0.010930603939676818

In [60]:
i2= solve_fun(pv, pmt2, n2, fv)
print(sol2)
i2

0.01063980717924384


0.01063980717901046

In [61]:
i3= solve_fun(pv, pmt3, n3, fv)
print(sol3)
i3

0.010697698792848072


0.010697698792974464

In [62]:
iy1 = 12*i1
iy2 = 12*i2
iy3 = 12*i3
print("I1year=", iy1*100,'%')
print("I2year=", iy2*100,'%')
print("I3year=", iy3*100,'%')

I1year= 13.11672472761218 %
I2year= 12.767768614812553 %
I3year= 12.837238551569358 %


# Отчёт

Функция, решающая уравнение $ PV*(1+i)^n + PMT*\frac{(1+i)^n-1}{i} + FV = 0 $ (1) выглядит следующим образом: 
```
def solve_fun(pv, pmt, n, fv):
    def fun(i):
        return pv*(1+i)**n + pmt*((((1+i)**n)-1)/i) + fv
    return brentq(fun, 0.001, 1.0)
```
Для решения уравнения (1) был исрользован метод Брента как наиболее популярный и универсальный.

2. Годовые ставки составили:
- 13.12% для примера 1
- 12.77% для примера 2
- 12.84% для примера 3

Из 3х представленых вариантов наиболее привлекательным является 2ой вариант из-за наименьшей ставки.